### PR TITLE
[FLINK-32023][API / DataStream] Support negative Duration for special usage, such as -1ms for execution.buffer-timeout

### PR DESCRIPTION
**Pull request by Myracle**
## What is the purpose of the change

*Support negative Duration for special usage, such as -1ms for execution.buffer-timeout*

## Brief change log

  - *Support to parse the negative sign in TimeUtils's method parseDuration.*

## Verifying this change
This change added tests and can be verified as follows:
  - *Added test testParseDurationWithNegativeNumber in TimeUtilsTest.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/util/TimeUtils.java [hunk 1]
```diff
@@ -69,7 +69,8 @@ public static Duration parseDuration(String text) {
         int pos = 0;
 
         char current;
-        while (pos < len && (current = trimmed.charAt(pos)) >= '0' && current <= '9') {
+        // Negative number is supported for special usage, such as -1ms for execution.buffer-timeout
+        while (pos < len && ((current = trimmed.charAt(pos)) >= '0' && current <= '9' || current == '-')) {
```
Comment (1996fanrui) [inline flink-core/src/main/java/org/apache/flink/util/TimeUtils.java:73]: We should check that the `-` must be the first char, right?

File: flink-core/src/test/java/org/apache/flink/util/TimeUtilsTest.java [hunk 2]
```diff
@@ -118,6 +118,12 @@ public void testParseDurationTrim() {
         assertEquals(155L, TimeUtils.parseDuration("      155      ms   ").toMillis());
     }
 
+    @Test
+    public void testParseDurationWithNegativeNumber() {
+        assertEquals(-1L, TimeUtils.parseDuration("-1ms").toMillis());
+        assertEquals(-1L, TimeUtils.parseDuration("-1").toMillis());
```
Comment (1996fanrui) [inline flink-core/src/test/java/org/apache/flink/util/TimeUtilsTest.java:124]: Could you add some cases with multiple `-`?

**Review by 1996fanrui — COMMENTED** [review]
Hi @Myracle , thanks for your contribution.

I see the CI fails, please run `mvn spotless:apply` before commiting.

**Comment by zentol** [comment]
Could we not do this? This now break assumptions that _every_ other component can make in that durations are always positive.

Either use Configuration#getOptional (+ adding it by default to flink-conf.yaml) or add a boolean option to disable the timeout.

**Comment by Myracle** [comment]
@zentol I agree with you for that the change may affect other usages. What about adding a boolean config named execution.buffer-timeout.enabled which default value is true. When it is false, flushing only when the output buffer is full. At the same time, the execution.buffer-timeout's value will be ignored. cc @1996fanrui

**Comment by 1996fanrui** [comment]
Thanks @zentol 's feedback here.

> What about adding a boolean config named execution.buffer-timeout.enabled which default value is true. When it is false, flushing only when the output buffer is full. At the same time, the execution.buffer-timeout's value will be ignored. 

Sounds make sense.

File: flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java [hunk 3]
```diff
@@ -92,6 +92,17 @@ public class ExecutionOptions {
                     .withDescription(
                             "Tells if we should use compression for the state snapshot data or not");
 
+    public static final ConfigOption<Boolean> BUFFER_TIMEOUT_ENABLED =
+            ConfigOptions.key("execution.buffer-timeout.enabled")
```
Comment (1996fanrui) [inline flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:96]: FLINK-29372 limited that the option key cannot be a prefix of other options, however `execution.buffer-timeout` is the prefix of `execution.buffer-timeout.enabled`. So the CI fails.

Hi @zentol , could we update the option key for `BUFFER_TIMEOUT` and mark the `"execution.buffer-timeout"` as the DeprecatedKey? How about update it to `execution.buffer-timeout.interval` ? Or do you have any suggestions here? Looking forward to your opinion.

Comment (Myracle) [reply flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:96]: How about changing the config execution.buffer-timeout.enabled to execution.flush-on-buffer-full?

Comment (1996fanrui) [reply flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:96]: Hi @Myracle , thanks for your feedback.

I see too many related option names that all have the same prefix but different suffixes. From the user's point of view, the same prefix also clearly tells the user that these options are related. Therefore, when FLINK-29372 restricts the option prefix, it also adds suffixes to many option names that do not conform to the specification.

So I prefer keep the prefixes the same for these 2 options. In order to the reasonable of this fix, I look forward to more feedback from community.

Hi @wanglijie95  @reswqa , would you mind take a look this PR in your free time? thanks a lot.

Comment (reswqa) [reply flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:96]: Rename it to `execution.buffer-timeout.interval` and mark the old as deprecated key is ok from my side. But I want to hear @zentol's thoughts.

Comment (reswqa) [reply flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:96]: BTW, I'm not sure do we need to start a discussion about this changes of configuration in the mailing list.

Comment (wanglijie95) [reply flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:96]: +1 for name `execution.buffer-timeout.interval`

Comment (Myracle) [reply flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:96]: Rename it to execution.buffer-timeout.interval. @zentol Would you like to give your thought? Thanks.

Comment (zentol) [reply flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:96]: `execution.buffer-timeout.interval` and `execution.buffer-timeout.enabled` sound fine to me.

File: flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java [hunk 4]
```diff
@@ -92,6 +92,43 @@ public class ExecutionOptions {
...
+                                                            + BUFFER_TIMEOUT_ENABLED.key()
+                                                            + " is false,"
+                                                            + " trigger flushing only when the output buffer is full thus maximizing "
+                                                            + "throughput"))
+                                    .build());
+
+    /** @deprecated Use {@link #BUFFER_TIMEOUT_INTERVAL} instead. */
+    @Deprecated
     public static final ConfigOption<Duration> BUFFER_TIMEOUT =
             ConfigOptions.key("execution.buffer-timeout")
```
Comment (1996fanrui) [inline flink-core/src/main/java/org/apache/flink/configuration/ExecutionOptions.java:133]: Don't need to create a `ConfigOption`, you can take a look this [PR](https://github.com/apache/flink/pull/20867/files#diff-7d4e333f055cb785073ea9bf5a53fcc7f8f527ba4230e153dd9e59212093aec4).

image

**Review by 1996fanrui — CHANGES_REQUESTED** [review]

**Review by 1996fanrui — APPROVED** [review]
Thanks @Myracle 's update, LGTM.

**Comment by 1996fanrui** [comment]
I will merge this PR next Monday if no other comments.

**Comment by 1996fanrui** [comment]
I will merge this PR next Monday if no other comments.

Hi @Myracle , could you help squash all commits? thanks~

**Comment by 1996fanrui** [comment]
@flinkbot run azure

**Comment by 1996fanrui** [comment]
Thanks everyone who discussed here, and @Myracle 's contribution, merging.

**Pull request merged**